runtime/src/snapshot_minimizer.rs

```diff
@@ -0,0 +1,609 @@
+use {
+    crate::{
+        accounts_db::{AccountStorageEntry, AccountsDb, PurgeStats},
+        bank::Bank,
+    },
+    dashmap::DashSet,
+    log::info,
+    rayon::{
+        iter::{IntoParallelIterator, IntoParallelRefIterator, ParallelIterator},
+        prelude::ParallelSlice,
+    },
+    solana_measure::measure,
+    solana_sdk::{
+        account::ReadableAccount,
+        account_utils::StateMut,
+        bpf_loader_upgradeable::{self, UpgradeableLoaderState},
+        clock::Slot,
+        pubkey::Pubkey,
+    },
+    std::{
+        collections::HashSet,
+        sync::{
+            atomic::{AtomicUsize, Ordering},
+            Arc, Mutex,
+        },
+    },
+};
+
+pub struct SnapshotMinimizer {
+    bank: Arc<Bank>,
+    starting_slot: Slot,
+    ending_slot: Slot,
+    minimized_account_set: DashSet<Pubkey>,
+}
+
+impl SnapshotMinimizer {
+    /// Removes all accounts not necessary for replaying slots in the range [starting_slot, ending_slot].
+    pub fn minimize(
+        bank: Arc<Bank>,
+        starting_slot: Slot,
+        ending_slot: Slot,
+        minimized_account_set: DashSet<Pubkey>,
+    ) {
+        let minimizer = Self::new(bank, starting_slot, ending_slot, minimized_account_set);
+
+        minimizer.add_accounts(
+            Self::get_rent_collection_accounts,
+            "rent collection accounts",
+        );
+        minimizer.add_accounts(Self::get_vote_accounts, "vote accounts");
+        minimizer.add_accounts(Self::get_stake_accounts, "stake accounts");
+        minimizer.add_accounts(Self::get_owner_accounts, "owner accounts");
+        minimizer.add_accounts(Self::get_programdata_accounts, "programdata accounts");
+
+        minimizer.minimize_accounts_db();
+
+        // Update accounts_cache and capitalization
+        minimizer.bank.force_flush_accounts_cache();
+        minimizer.bank.set_capitalization();
+    }
+
+    /// Create a minimizer - used in tests
+    fn new(
+        bank: Arc<Bank>,
+        starting_slot: Slot,
+        ending_slot: Slot,
+        minimized_account_set: DashSet<Pubkey>,
+    ) -> Self {
+        Self {
+            bank,
+            starting_slot,
+            ending_slot,
+            minimized_account_set,
+        }
+    }
+
+    /// Helper function to measure time and number of accounts added
+    fn add_accounts<F>(&self, add_accounts_fn: F, name: &'static str)
+    where
+        F: Fn(&SnapshotMinimizer),
+    {
+        let initial_accounts_len = self.minimized_account_set.len();
+        let (_, measure) = measure!(add_accounts_fn(self), name);
+        let total_accounts_len = self.minimized_account_set.len();
+        let added_accounts = total_accounts_len - initial_accounts_len;
+
+        info!(
+            "Added {added_accounts} {name} for total of {total_accounts_len} accounts. get {measure}"
+        );
+    }
+
+    /// Used to get rent collection accounts in `minimize`
```

Comment:
add comment:

/// related to Bank::rent_collection_partitions()

Reply:
also add to comment something like:

/// add all pubkeys we would collect rent from or rewrite to 'minimized_account_set'

Reply:
Added more descriptive descriptions for all the functions in this file